runtime/drivers/github/github.go

```diff
@@ -137,38 +124,40 @@ func (c *connection) MigrationStatus(ctx context.Context) (current, desired int,
 	return 0, 0, nil
 }
 
-// clone runs the initial clone of the repo.
-func (c *connection) clone(ctx context.Context) error {
-	cloneURL, err := c.cloneURL(ctx)
-	if err != nil {
-		return err
+// pull pulls changes from the repo. It also clones the repo first time.
+func (c *connection) pull(ctx context.Context) error {
+	cloneTried := false
+	var cloneErr error
+	if !c.cloned.Load() {
+		func() {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+			// check again in case cloned already by concurrent request
+			if !c.cloned.Load() {
+				cloneTried = true
+				cloneErr = c.clone(ctx)
+				if cloneErr == nil {
+					c.cloned.Store(true)
+				}
+			}
+		}()
 	}
-
-	_, err = git.PlainClone(c.tempdir, false, &git.CloneOptions{
-		URL:           cloneURL,
-		ReferenceName: plumbing.NewBranchReferenceName(c.dsn.Branch),
-		SingleBranch:  true,
-	})
-	if err != nil {
-		return err
+	// no need to pull again if clone tried within same request
+	if cloneTried {
+		return cloneErr
```

Comment:
Not sure what this is supposed to do, but pretty sure it doesn't work. The goroutine will likely get scheduled after the `if` check and it will try to pull instead of clone.

I don't think we need to overthink this, just use a lock on the whole `pull` function. If you want to deduplicate requests, use a `TryLock` first and if it fails, you can do a normal `Lock` and return immediately when you get it (maybe there's an even simpler way).

Reply:
> Not sure what this is supposed to do, but pretty sure it doesn't work. The goroutine will likely get scheduled after the if check and it will try to pull instead of clone.

I think you misunderstood. The clone happens in the same goroutine. 

The expectation is that only the first request should clone the project and subsequent requests pull. 
Even if we take lock on the whole pull function, we would still need to keep a boolean flag to denote that its been cloned already so it really makes sense to just use an atomic boolean and check its value instead. 
I can remove that `cloneTried` thing and put all of this logic in the clone function. Here is how the impl would like (Although this will both clone and pull for the first request which should be fine I guess) :


```

func (c *connection) pull(ctx context.Context) error {
	if !c.cloned.Load() {
		if err := c.clone(ctx); err != nil {
			return err
		}
	}
	**** pull logic ***
}

// clone runs the initial clone of the repo.
func (c *connection) clone(ctx context.Context) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.cloned.Load() {
		// cloned already
		return nil
	}

	cloneURL, err := c.cloneURL(ctx)
	if err != nil {
		return err
	}

	_, err = git.PlainClone(c.tempdir, false, &git.CloneOptions{
		URL:           cloneURL,
		ReferenceName: plumbing.NewBranchReferenceName(c.dsn.Branch),
		SingleBranch:  true,
	})
	if err != nil {
		return err
	}

	c.cloned.Store(true)
	return nil
}

```

Reply:
I see now there's no `go` call, but it still doesn't really make sense. There is no need here for anything fancy apart from locking the whole call (it's just as important to have a lock around the pull as it is for the clone), and maybe doing a `TryLock` to deduplicate calls as described earlier.

Reply:
The reason there is a lock on the clone is that we don't want clone to run multiple times. It wasn't required previously because it was happening on `open` call and there was no chance for it to run concurrently. 
For pull I didn't add any lock since it wasn't there previously as well. I think you are right here. We need to also lock around pull. There is a chance that it can leave the local clone in a weird state. 

I don't need understand the use of `TryLock`. For clone we just don't want `deduplication` we only want it to run just once(successfully) during the complete lifecycle.

Reply:
I mean it can just be something like this:

```go
func (c *connection) pull(ctx context.Context) error {
  var deduplicated bool
  if !c.mu.TryLock() {
    deduplicated = true
    c.mu.Lock()
  }
  defer c.mu.Unlock()

  if deduplicated {
    return nil
  }

  if !c.cloned {
    // do clone
    return nil
  }

  // do pull
  return nil
}
```

Reply:
To return deduplicated errors, you can just store the last pull err as a var

Reply:
You can also use a singleflight for it with a const key if you want